src/helix/core.clj

```diff
@@ -72,11 +72,13 @@
 
     (provider {:context my-context :value my-value} child1 child2 ...childN)"
   [{:keys [context value] :as props} & children]
-  `^js/React.Element ($ (.-Provider ~context)
-                        ;; use contains to guard against `nil`
-                        ~@(when (contains? props :value)
-                            `({:value ~value}))
-                        ~@children))
+  (let [ctx (vary-meta (gensym "ctx") assoc :tag 'js)]
+    `(let [~ctx ~context]
+       ^js/React.Element ($ (.-Provider ~ctx)
```

Comment:
Could we emit `^js` here instead?

Reply:
How would that work?  I was going by [this answer](https://stackoverflow.com/questions/11919602/generating-clojure-code-with-macro-containing-type-hints) which seems to imply this is the only way to get the desired result in a macro.  It also implies that the `^js/React.Element` hint may be being applied to the form the macro is evaluating, not the form the macro is emitting.

Reply:
See also https://clojureverse.org/t/clojurescript-macros-and-type-hints/7292

Reply:
@lilactown Yes, `^js` can be used instead of `^js/React.Element`. The extra type info is for readability, but shadow-cljs for example actually throws it away.  I think Closure conservatively refuses to mangle recognized property names regardless of the object type.  I’m looking to document this soon on the cljs api site, see this issue for more details: https://github.com/cljs/api/issues/189

Reply:
@shaunlebron my question was actually about the line

```clojure
  (let [ctx (vary-meta (gensym "ctx") assoc :tag 'js)]
```

which is associng the tag to the symbol's meta instead of emitting a reader literal to assign the meta in the quoted form.

Reply:
I don’t think so, because the syntax quote fully qualifies inner symbols, turning `^js` into `^<calling-ns>/js`:

```clj
(:tag (meta `^js (.-hello world)) ;; => cljs.user/js
```

But this is only the macro-expanded form.  After the result is evaluated, the metadata tag symbol is evaluated (only if the form it belongs to is evaluated, which is the case for the invocation `(.-hello world)`)— whereas we want the *literal* symbol, not evaluated.

This makes my first comment here wrong. And now, I don’t actually know if the `^js/React.Element` is working.  I’ve asked the compiler team on slack.  I might’ve missed something, but I’ll let you know what they say.

Reply:
You can’t do `^js ~ctx` in a syntax quote, but `^{:tag ~'js} ~ctx` works.

`^js/React.Context ~ctx` or `^js/Object ~ctx` are also fine.

@thheller recommends `(defn provider [^js ctx] (.-Provider ctx))`

Reply:
I don't think `^js/React.Context ~ctx` or `^js/Object ~ctx` work inside a syntax quote.  See:

https://stackoverflow.com/questions/11919602/generating-clojure-code-with-macro-containing-type-hints
https://clojureverse.org/t/clojurescript-macros-and-type-hints/7292

Reply:
Alternatively, I just verified that this works:

```clj
`(let [^{:tag ~'js} ctx# ~context]
     ^js/React.Element (helix.core/$ (.-Provider ctx#)
...
```

But #109 does not work.

Reply:
how does it work for `^js/React.Element` but not `^js/React.Context`?

Reply:
I don't think it works for `^js/React.Element` either, but just in my case that's not causing problems.